### ISSUE TITLE
Exception in hamt-ref: Key is not in the hamt with irritant 13982459685072098939

### ISSUE THREAD

**Issue by amirouche** [ijp/pfds#5]
Using chez scheme 9.5 and the following program:

```scheme
(call-with-values (lambda () (let loop ((keys '())
                                        (hamt (make-hamt values equal?))
                                        (count 0))
                               (if (equal? count 10)
                                   (values keys hamt)
                                   (let ((value (random (expt 2 64))))
                                     (loop (cons value keys) (hamt-set hamt value #t) (+ 1 count))))))
  (lambda (keys hamt)
    (let loop ((keys keys))
      (unless (null? keys)
        (hamt-ref hamt (car keys))
        (loop (cdr keys))))))
```

**Comment by amirouche** [comment]
If I replace `values` with srfi 69 `hash` function it leads to the same behavior.

**Comment by amirouche** [comment]
It's prolly because I used a wrong hash function.

**Comment by pukkamustard** [comment]
I'm having the same problem. I agree that it is because the hash function does not satisfy some property. But I haven't figured out what that property is.

@amirouche What hash function did you end up using?

**Comment by amirouche** [comment]
@pukkamustard did you try https://github.com/ijp/pfds/pull/6

**Comment by amirouche** [comment]
I do not remember what was the fix.

**Comment by amirouche** [comment]
Try the fix is #6 it seems related.

**Comment by pukkamustard** [comment]
That looks very related. Thanks for the pointer!

Will try and report back.

**Comment by pukkamustard** [comment]
Unfortunately, the bug seems to persist even with the fixes in #6...

**Comment by amirouche** [comment]
In babelia repo, master branch, inside `scheme/mapping/hash` there is a similar datastructure https://github.com/amirouche/babelia/tree/master/scheme/mapping

Otherwise look into Andy's `fash`.
